File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    answer = []
    genre_dict = {}
    play_count = {}
    
    for i in range(len(genres)):
        genre = genres[i]
        play = plays[i]
        if genre not in genre_dict:
            genre_dict[genre] = []
            play_count[genre] = 0
        genre_dict[genre].append((i, play))
        play_count[genre] += play
        
    sort_genres = sorted(play_count.items(), key = lambda x: x[1], reverse = True)
    
    for genre, _ in sort_genres:
        sort_songs = sorted(genre_dict[genre], key = lambda x: (-x[1], x[0]))
        answer.extend([i for i,_ in sort_songs[:2]])
        
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.js
function solution(genres, plays) {
    const dic = {};
    const dupDic = {};
    genres.forEach((g,i)=> dic[g] = dic[g] ?  dic[g] + plays[i] : plays[i]);

    return genres          
          .map((g,i)=> ({genre: g, count: plays[i], index: i}))
          .sort((a,b)=>{               
               if (a.genre !== b.genre) return dic[b.genre] - dic[a.genre];
               if (a.count !== b.count) return b.count - a.count;
               return a.index - b.index;
           })
           .filter(v => {
               if (dupDic[v.genre] >= 2) return false;
               dupDic[v.genre] = dupDic[v.genre] ? dupDic[v.genre] + 1 : 1;
               return true;
            })
           .map(v => v.index);    
}